### ISSUE TITLE
Make Yices2 available as an available-solver

### ISSUE THREAD

**Issue by pjljvandelaar** [TorXakis/TorXakis#265]
Add yices 2 solver-id, with the appropriate executable name and flags in the 
`.torxakis.yaml` file.
Yices can be obtained from https://github.com/SRI-CSL/yices2 under the GPL license.

**Comment by pjljvandelaar** [comment]
benefits: 
* performance (when Yices outperforms other solvers ;-)
* robustness (test TorXakis with more smt solvers)
* reduce dependency on z3 and cvc4

**Comment by pjljvandelaar** [comment]
Watch https://github.com/SRI-CSL/yices2/issues/77: if license is fine we can continue

**Comment by capitanbatata** [comment]
Then we can remove the blocked label, can't we?

**Comment by pjljvandelaar** [comment]
On 09/29/2017 02:37 AM, Pierre_van_der_Laar (Functional Account) wrote:
> Hi Bruno,
> 
> Thanks for the info.
> Do you agree that
> since Yices2 is independently installed and TorXakis calls Yices2 via a cmdline interface,
> everything is fine: We can use Yices2 and we will keep our BSD3 license!
> 
> Greetings,
>     Pierre

That's fine by me, but I'm not a lawyer.

Bruno

> 
> 
> -----Original Message-----
> From: Bruno Dutertre [mailto:bruno.dutertre@sri.com]
> Sent: 2017 Sep 28 6:21 PM
> To: Pierre_van_der_Laar (Functional Account); fm-licensing@csl.sri.com
> Subject: Re: License question
> 
> On 9/28/17 6:48 AM, Pierre_van_der_Laar (Functional Account) wrote:
>> Dear Yices 2 support,
>>
>> We are developing TorXakis under the BSD3 license (see
>> https://github.com/TorXakis/TorXakis for more info).
>>
>> TorXakis uses a problem solver under the hood.
>>
>> Currently, we test TorXakis with cvc4 and z3 (and z3str3).
>>
>> We would like to test TorXakis with more problem solvers including Yices 2.
>>
>> Based on the license text we have doubt whether this usage is allowed.
>>
>> Could you please clarify whether we are allowed to download Yices 2
>>
>> and test TorXakis with Yices 2?
>>
>> Of course, Yices 2 will not be distributed with TorXakis.
>>
> 
> Yices 2 is now open source (GPL). The latest release is on github:
> https://github.com/SRI-CSL/yices2.
> 
> Hope that satisfies your need,
> 
> Bruno
> 
>> Greetings,
>>
>> Pierre van de Laar
>>
>> Research Fellow
>>
>> TNO-ESI
>>
>>
>> ------------------------------------------------------------------------
>> The information contained in this message may be confidential and
>> legally protected under applicable law. The message is intended solely
>> for the addressee(s). If you are not the intended recipient, you are
>> hereby notified that any use, forwarding, dissemination, or reproduction
>> of this message is strictly prohibited and may be unlawful. If you are
>> not the intended recipient, please contact the sender by return e-mail
>> and destroy all copies of the original message.
> 
> 
> 
> ________________________________
> The information contained in this message may be confidential and legally protected under applicable law. The message is intended solely for the addressee(s). If you are not the intended recipient, you are hereby notified that any use, forwarding, dissemination, or reproduction of this message is strictly prohibited and may be unlawful. If you are not the intended recipient, please contact the sender by return e-mail and destroy all copies of the original message.
>

**Comment by pjljvandelaar** [comment]
Based on the mail above I think we can remove the blocked label.

I haven't seen a response on the issue SRI-CSL/yices2#77SRI-CSL/yices2#77, GPL with linking exception or LGPL.

**Comment by pjljvandelaar** [comment]
Latest version can be downloaded from  http://yices.csl.sri.com/
(see answer on https://github.com/SRI-CSL/yices2/issues/78)

**Comment by capitanbatata** [comment]
I don't think any of the use cases you mention in https://github.com/SRI-CSL/yices2/issues/77 can be considered linking, if we use the software definition of it. I would say let's just use the tool.

Regarding downloading, in Unix systems `z3`, `cvc4, and `yices` can be installed via [`nix`](https://nixos.org/nix/).

**Comment by pjljvandelaar** [comment]
currently locked due to Yices not recognizing the `ALL` logic:
https://github.com/SRI-CSL/yices2/issues/84

**Comment by pjljvandelaar** [comment]
Even the simple `Echo.txs` examples fail
```
C:\TorXakisSandbox\TorXakis.git>torxakis 59926 examps\Echo\Echo.txs

TXS >>  TorXakis :: Model-Based Testing

TXS >>  txsserver starting: "PC-16153.tsn.tno.nl" : 59926
TXS >>  Solver "yices" initialized : Yices [2.5.4]
TXS >>  TxsCore initialized
TXS >>  input files parsed:
TXS >>  examps\Echo\Echo.txs
TXS >> stepper Model
TXS >>  Stepper started
TXS >> step 100
torxakis: <socket: 572>: hGetLine: does not exist (No such file or directory)
```

We generated the following smt:
```
(get-info :name)
(get-info :version)
(set-logic ALL)
(set-option :produce-models true)
(set-info :smt-lib-version 2.5)
(push 1)
(push 1)
(declare-fun $In$1012$0$1$() Int)
(check-sat)
```

Yices produces as output
```
(:name "Yices")
(:version "2.5.4")
(error "unknown logic: ALL")
(error "no logic set")
(error "no logic set")
(error "at line 17, column 30: undefined sort: Int")
```

**Comment by pjljvandelaar** [comment]
And additionally: https://github.com/SRI-CSL/yices2/issues/85

**Comment by pjljvandelaar** [comment]
The current version is Yices 2.6.0. It was released on 2018-06-30.

**Comment by pjljvandelaar** [comment]
If yices works add it to the ini-file.
Also check yices nl since non-linear solving is really relevant for one of our customers/partners.

**Comment by pjljvandelaar** [comment]
Made a change since Yices is more strict (see #756).

Yices does not support recursive functions: going submit issue!
So we are blocked once more ;-(

**Comment by pjljvandelaar** [comment]
We can already offer it when user **don't** use 
* functions (since we map them on recursive functions) 
* data types - not yet supported by Yices
* strings - not yet supported by Yices

@tretmans  Is that sufficient?

Note: we can't distribute Yices together with TorXakis due to its license. So we have to make our users aware of **MANY** constraints and download/installation procedures.

**Comment by pjljvandelaar** [comment]
Solved in #758 
Yet with the constraints:  No functions, no data types and no strings!
